File: java/1438_longestSubarray.java
// TreeMap
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        if(nums.length==0)
            return 0;
        int p1=0,p2=0,res=0;
        TreeMap<Integer,Integer> t = new TreeMap<>();
        while(p2<nums.length)
        {
            if(t.get(nums[p2])!=null)
                t.put(nums[p2],1+t.get(nums[p2]));
            else
                t.put(nums[p2],1);
            while(p1<=p2 && t.lastKey()-t.firstKey()>limit)
            {
                t.put(nums[p1],t.get(nums[p1])-1);
                if(t.get(nums[p1])==0)
                    t.remove(nums[p1]);
                p1++;
            }
            res = Math.max(res,++p2-p1);
        }
        return res;
    }
}

File: cpp/1438_longestSubarray.cpp
// 连续子数组，无脑sliding window
// 右指针主动右移，每次用一个key有序的container来存储出现次数
// 当container内key的极差大于limit时，左指针右移，同时更新container
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        if(nums.size()==0)
            return 0;
        int p1=0,p2=0,res=0;
        map<int,int> m;
        while(p2<nums.size())
        {
            m[nums[p2]]++;
            while(p1<=p2 && m.rbegin()->first-m.begin()->first>limit)
            {
                m[nums[p1]]--;
                if(m[nums[p1]]==0)
                    m.erase(nums[p1]);
                p1++;
            }
            res = max(res,p2-p1+1);
            p2++;
        }
        return res;
    }
};
